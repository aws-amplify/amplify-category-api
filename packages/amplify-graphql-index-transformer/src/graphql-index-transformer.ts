import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InvalidDirectiveError,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { IndexDirective } from '@aws-amplify/graphql-directives';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isListType, isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { generateKeyAndQueryNameForConfig, validateNotSelfReferencing } from './utils';

/**
 *
 */
export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];

  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', IndexDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as IndexDirectiveConfiguration,
      generateGetArgumentsInput(context.transformParameters),
    );

    /**
     * Impute Optional Fields
     * Start with sort key fields since both the name and queryField will work sort key fields into the generated name if necessary.
     */
    args.sortKeyFields = getOrGenerateDefaultSortKeyFields(args);
    args.name = getOrGenerateDefaultName(args);
    args.queryField = getOrGenerateDefaultQueryField(context, args);
    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

/**
 * Return the name if provided in our args, else
 * compute the name based on the field name, and sortKeyFields.
 */
const getOrGenerateDefaultName = (config: IndexDirectiveConfiguration): string => {
  const indexNameRegex = /^[A-Za-z0-9_\-\.]{3,255}$/;
  if (config.name) {
    if (!indexNameRegex.test(config.name)) {
      throw new Error(
        `The indexName is invalid. It should be between 3 and 255 characters. Only A–Z, a–z, 0–9, underscore characters, hyphens, and periods allowed.`,
      );
    }
    return config.name;
  }

  if (config.name === null) {
    throw new Error('Explicit null value not allowed for name field on @index');
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * Return the queryField if provided in our args, else
 * compute the queryField based on the field name, and sortKeyFields if the feature flag is enabled.
 */
const getOrGenerateDefaultQueryField = (
  context: TransformerSchemaVisitStepContextProvider,
  config: IndexDirectiveConfiguration,
): string | null => {
  const autoIndexQueryNamesIsEnabled = context.transformParameters.enableAutoIndexQueryNames;
  // Any explicit null will take effect, if enableAutoIndexQueryNames and no queryField is provide set to null for consistency
  if (config.queryField === null || (!autoIndexQueryNamesIsEnabled && !config.queryField)) {
    return null;
  }

  if (config.queryField) {
    return config.queryField;
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * sortKeyFields are optional so default to empty list,
 * if we get a raw object just wrap in an array,
 * else return the list which was provided correctly.
 */
const getOrGenerateDefaultSortKeyFields = (config: IndexDirectiveConfiguration): string[] => {
  if (!config.sortKeyFields) {
    return [];
  }
  if (!Array.isArray(config.sortKeyFields)) {
    return [config.sortKeyFields];
  }
  return config.sortKeyFields;
};

const validate = (config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void => {
  const { name, object, field, sortKeyFields } = config;
  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find((directive) => directive.name.value === 'model');

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${IndexDirective.name} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        const hasSortFields = peerDirective.arguments!.some(
          (arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0,
        );
        config.primaryKeyField = objectField;

        if (!hasSortFields && objectField.name.value === field.name.value) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === IndexDirective.name &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${IndexDirective.name} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }

    for (const peerDirective of objectField.directives!) {
      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0);

      if (
        !ctx.transformParameters.secondaryKeyAsGSI &&
        !hasSortFields &&
        objectField == config.primaryKeyField &&
        objectField.name.value === field.name.value
      ) {
        throw new InvalidDirectiveError(
          `Invalid @index '${name}'. You may not create an index where the partition key ` +
            'is the same as that of the primary key unless the index has a sort field. ' +
            'You cannot have a local secondary index without a sort key in the index.',
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums) || isListType(field.type)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums) || isListType(sortField.type)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
};
