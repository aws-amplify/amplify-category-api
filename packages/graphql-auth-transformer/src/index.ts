export * from './ModelAuthTransformer';
// No-op change to trigger re-publish
