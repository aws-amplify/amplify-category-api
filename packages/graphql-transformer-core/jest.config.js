const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
