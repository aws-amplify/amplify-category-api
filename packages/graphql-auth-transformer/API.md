## API Report File for "graphql-auth-transformer"

> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).

```ts

import { CompoundExpressionNode } from 'graphql-mapping-template';
import { default as default_2 } from 'cloudform-types/types/appSync/apiKey';
import { default as default_3 } from 'cloudform-types/types/appSync/resolver';
import { default as default_4 } from 'cloudform-types/types/appSync/dataSource';
import { DirectiveNode } from 'graphql';
import { Expression } from 'graphql-mapping-template';
import { FieldDefinitionNode } from 'graphql';
import GraphQLApi from 'cloudform-types/types/appSync/graphQlApi';
import { InterfaceTypeDefinitionNode } from 'graphql';
import ManagedPolicy from 'cloudform-types/types/iam/managedPolicy';
import { NumberParameter } from 'cloudform-types';
import { ObjectTypeDefinitionNode } from 'graphql';
import { StringParameter } from 'cloudform-types';
import Template from 'cloudform-types/types/template';
import { Transformer as Transformer_2 } from 'graphql-transformer-core';
import { TransformerContext } from 'graphql-transformer-core';

// @public (undocumented)
export type ApiKeyConfig = {
    description?: string;
    apiKeyExpirationDays: number;
    apiKeyExpirationDate?: Date;
};

// @public (undocumented)
export type AppSyncAuthConfiguration = {
    defaultAuthentication: AppSyncAuthConfigurationEntry;
    additionalAuthenticationProviders: Array<AppSyncAuthConfigurationEntry>;
};

// @public (undocumented)
export type AppSyncAuthConfigurationEntry = {
    authenticationType: AppSyncAuthMode;
    apiKeyConfig?: ApiKeyConfig;
    userPoolConfig?: UserPoolConfig;
    openIDConnectConfig?: OpenIDConnectConfig;
    lambdaAuthorizerConfig?: LambdaAuthorizerConfig;
};

// @public (undocumented)
export type AppSyncAuthMode = 'API_KEY' | 'AMAZON_COGNITO_USER_POOLS' | 'AWS_IAM' | 'OPENID_CONNECT' | 'AWS_LAMBDA';

// @public (undocumented)
export type ConfiguredAuthProviders = {
    default: AuthProvider;
    onlyDefaultAuthProviderConfigured: boolean;
    hasApiKey: boolean;
    hasUserPools: boolean;
    hasOIDC: boolean;
    hasIAM: boolean;
};

// @public (undocumented)
export type LambdaAuthorizerConfig = {
    lambdaFunction: string;
    ttlSeconds?: number;
};

// @public (undocumented)
export class ModelAuthTransformer extends Transformer_2 {
    constructor(config?: ModelAuthTransformerConfig);
    // (undocumented)
    after: (ctx: TransformerContext) => void;
    // (undocumented)
    authPolicyResources: Set<string>;
    // (undocumented)
    before: (ctx: TransformerContext) => void;
    // (undocumented)
    config: ModelAuthTransformerConfig;
    // (undocumented)
    configuredAuthProviders: ConfiguredAuthProviders;
    // (undocumented)
    field: (parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode, definition: FieldDefinitionNode, directive: DirectiveNode, ctx: TransformerContext) => void;
    // (undocumented)
    generateIAMPolicyforAuthRole: boolean;
    // (undocumented)
    generateIAMPolicyforUnauthRole: boolean;
    // (undocumented)
    hasProviderAuthRules(rules: AuthRule[]): Boolean;
    // (undocumented)
    object: (def: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerContext) => void;
    // Warning: (ae-forgotten-export) The symbol "AuthRule" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    protectSyncQuery(ctx: TransformerContext, def: ObjectTypeDefinitionNode, resolverResourceID: string, rules: AuthRule[]): void;
    // Warning: (ae-forgotten-export) The symbol "ResourceFactory" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    resources: ResourceFactory;
    // (undocumented)
    unauthPolicyResources: Set<string>;
}

// @public (undocumented)
export type ModelAuthTransformerConfig = {
    authConfig?: AppSyncAuthConfiguration;
    addAwsIamAuthInOutputSchema?: boolean;
};

// @public (undocumented)
export type OpenIDConnectConfig = {
    name: string;
    issuerUrl: string;
    clientId?: string;
    iatTTL?: number;
    authTTL?: number;
};

// @public (undocumented)
export type UserPoolConfig = {
    userPoolId: string;
};

// Warnings were encountered during analysis:
//
// src/ModelAuthTransformer.ts:144:3 - (ae-forgotten-export) The symbol "AuthProvider" needs to be exported by the entry point index.d.ts

// (No @packageDocumentation comment for this package)

```
