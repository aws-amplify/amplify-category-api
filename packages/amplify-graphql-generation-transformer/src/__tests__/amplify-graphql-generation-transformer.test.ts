import { AuthTransformer } from '@aws-amplify/graphql-auth-transformer';
import { IndexTransformer, PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY, validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';
import { DeploymentResources, testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { parse } from 'graphql';
import { GenerationTransformer } from '..';
import { BelongsToTransformer, HasManyTransformer, HasOneTransformer } from '@aws-amplify/graphql-relational-transformer';

const todoModel = `
type Todo @model {
  content: String
  isDone: Boolean
}`;

test('generation route model list response type', () => {
  const queryName = 'makeTodos';
  const inputSchema = `
    ${todoModel}

    type Query {
        ${queryName}(description: String!): [Todo]
        @generation(
          aiModel: "Claude3Haiku",
          systemPrompt: "Make a list of todo items based on the description.",
          inferenceConfiguration: {
            maxTokens: 100,
            temperature: 0.5,
            topP: 1.0
          }
        )
        @auth(rules: [{ allow:  public, provider: iam}])
    }
  `;

  expect(() => transform(inputSchema)).toThrow(/Disallowed required field type/);
});

test('generation route scalar type', () => {
  const queryName = 'makeTodo';
  const inputSchema = `
    type Query {
        ${queryName}(description: String!): String
        @generation(
          aiModel: "Claude3Haiku",
          systemPrompt: "Make a string based on the description.",
        )
    }
  `;
  const out = transform(inputSchema);
  expect(out).toBeDefined();

  const resolverCode = getResolverResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverCode).toBeDefined();
  expect(resolverCode).toMatchSnapshot();

  const resolverFnCode = getResolverFnResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverFnCode).toBeDefined();
  expect(resolverFnCode).toMatchSnapshot();

  expect(out.schema).toBeDefined();
  expect(out.schema).toMatchSnapshot();

  const schema = parse(out.schema);
  validateModelSchema(schema);
});

test('generation route custom query', () => {
  const queryName = 'generateRecipe';
  const inputSchema = `
    type Recipe {
      name: String
      ingredients: [String]
      instructions: String
    }

    type Query {
        ${queryName}(description: String!): Recipe
        @generation(
          aiModel: "Claude3Haiku",
          systemPrompt: "You are a helpful assistant that generates recipes.",
        )
    }
  `;
  const out = transform(inputSchema);
  expect(out).toBeDefined();

  const resolverCode = getResolverResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverCode).toBeDefined();
  expect(resolverCode).toMatchSnapshot();

  const resolverFnCode = getResolverFnResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverFnCode).toBeDefined();
  expect(resolverFnCode).toMatchSnapshot();

  const schema = parse(out.schema);
  validateModelSchema(schema);
});

test('generation route model type with null timestamps', () => {
  const queryName = 'makeTodo';
  const inputSchema = `
    type Todo @model(timestamps: {createdAt: null, updatedAt: null}) {
      content: String
      isDone: Boolean
    }

    type Query {
        ${queryName}(description: String!): Todo
        @generation(
          aiModel: "Claude3Haiku",
          systemPrompt: "Make a string based on the description.",
        )
    }
  `;
  const out = transform(inputSchema);

  const resolverCode = getResolverResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverCode).toBeDefined();
  expect(resolverCode).toMatchSnapshot();

  const resolverFnCode = getResolverFnResource(queryName, out.rootStack.Resources)['Properties']['Code'];
  expect(resolverFnCode).toBeDefined();
  expect(resolverFnCode).toMatchSnapshot();

  const schema = parse(out.schema);
  validateModelSchema(schema);
});

test('generation route required model type required field', () => {
  const queryName = 'makeTodo';
  const inputSchema = `
    ${todoModel}

    type Query {
        ${queryName}(description: String!): Todo!
        @generation(
          aiModel: "Claude3Haiku",
          systemPrompt: "Make a string based on the description.",
        )
    }
  `;

  expect(() => transform(inputSchema)).toThrow(/Disallowed required field type/);
});

describe('generation route invalid inference configuration', () => {
  const maxTokens = 'inferenceConfiguration: { maxTokens: 0 }';
  const temperature = 'inferenceConfiguration: { temperature: 1.1 }';
  const topP = 'inferenceConfiguration: { topP: -2 }';

  const generationRoute = (invalidInferenceConfig: string): string => {
    return `
      type Query {
          generate(description: String!): String
          @generation(
            aiModel: "Claude3Haiku",
            systemPrompt: "Make a string based on the description.",
            ${invalidInferenceConfig}
          )
      }
    `;
  };

  test('maxTokens invalid', () => {
    expect(() => transform(generationRoute(maxTokens))).toThrow(
      '@generation directive maxTokens valid range: Minimum value of 1. Provided: 0',
    );
  });

  test('temperature invalid', () => {
    expect(() => transform(generationRoute(temperature))).toThrow(
      '@generation directive temperature valid range: Minimum value of 0. Maximum value of 1. Provided: 1.1',
    );
  });

  test('topP invalid', () => {
    expect(() => transform(generationRoute(topP))).toThrow(
      '@generation directive topP valid range: Minimum value of 0. Maximum value of 1. Provided: -2',
    );
  });
});

const getResolverResource = (queryName: string, resources?: Record<string, any>): Record<string, any> => {
  const resolverName = `Query${queryName}Resolver`;
  return resources?.[resolverName];
};

const getResolverFnResource = (queryName: string, resources?: Record<string, any>): Record<string, any> => {
  const capitalizedQueryName = queryName.charAt(0).toUpperCase() + queryName.slice(1);
  const resourcePrefix = `Query${capitalizedQueryName}DataResolverFn`;
  if (!resources) {
    fail('No resources found.');
  }
  const resource = Object.entries(resources).find(([key, _]) => {
    return key.startsWith(resourcePrefix);
  })?.[1];

  if (!resource) {
    fail(`Resource named with prefix ${resourcePrefix} not found.`);
  }
  return resource;
};

const defaultAuthConfig: AppSyncAuthConfiguration = {
  defaultAuthentication: {
    authenticationType: 'AWS_IAM',
  },
  additionalAuthenticationProviders: [],
};

function transform(inputSchema: string, authConfig: AppSyncAuthConfiguration = defaultAuthConfig): DeploymentResources {
  const modelTransformer = new ModelTransformer();
  const authTransformer = new AuthTransformer();
  const indexTransformer = new IndexTransformer();
  const hasOneTransformer = new HasOneTransformer();
  const belongsToTransformer = new BelongsToTransformer();
  const hasManyTransformer = new HasManyTransformer();

  const transformers = [
    modelTransformer,
    new PrimaryKeyTransformer(),
    indexTransformer,
    hasManyTransformer,
    hasOneTransformer,
    belongsToTransformer,
    new GenerationTransformer(),
    authTransformer,
  ];

  // const processed = new GraphQLTransform({ transformers }).preProcessSchema(parse(inputSchema));
  // console.log(print(processed))

  const out = testTransform({
    schema: inputSchema,
    authConfig,
    transformers,
    dataSourceStrategies: {
      Todo: DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY,
    },
  });

  return out;
}
