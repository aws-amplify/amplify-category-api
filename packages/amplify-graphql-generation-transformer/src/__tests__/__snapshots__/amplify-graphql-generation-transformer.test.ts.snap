// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generation route custom query 1`] = `
Object {
  "Fn::Join": Array [
    "",
    Array [
      "export const request = (ctx) => {
  ctx.stash.typeName = \\"Query\\";
  ctx.stash.fieldName = \\"generateRecipe\\";
  ctx.stash.conditions = [];
  ctx.stash.metadata = {};
  ctx.stash.metadata.dataSourceType = \\"HTTP\\";
  ctx.stash.metadata.apiId = \\"",
      Object {
        "Fn::GetAtt": Array [
          "GraphQLAPI",
          "ApiId",
        ],
      },
      "\\";
  ctx.stash.connectionAttributes = {};
  ctx.stash.endpoint = \\"https://bedrock-runtime.",
      Object {
        "Ref": "AWS::Region",
      },
      ".amazonaws.com\\";
  ctx.stash.authRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "authRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.unauthRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "unauthRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.adminRoles = [];
  return {};
}

export const response = (ctx) => {
  return ctx.prev.result;
};",
    ],
  ],
}
`;

exports[`generation route custom query 2`] = `
"export function request(ctx) {
    const toolConfig = {\\"tools\\":[{\\"toolSpec\\":{\\"name\\":\\"responseType\\",\\"description\\":\\"Generate a response type for the given field\\",\\"inputSchema\\":{\\"json\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"name\\":{\\"type\\":\\"string\\",\\"description\\":\\"A UTF-8 character sequence.\\"},\\"ingredients\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"string\\",\\"description\\":\\"A UTF-8 character sequence.\\"}},\\"instructions\\":{\\"type\\":\\"string\\",\\"description\\":\\"A UTF-8 character sequence.\\"}},\\"required\\":[]}},\\"required\\":[\\"value\\"]}}}}],\\"toolChoice\\":{\\"tool\\":{\\"name\\":\\"responseType\\"}}};
    const prompt = \`You are a helpful assistant that generates recipes.\`;
    const args = JSON.stringify(ctx.args);

    return {
      resourcePath: '/model/anthropic.claude-3-haiku-20240307-v1:0/converse',
      method: 'POST',
      params: {
        headers: {
          'Content-Type': 'application/json',
        },
        body: {
          messages: [
          {
            role: 'user',
            content: [
              {
                text: args
              }
            ],
          }
        ],
        system: [{ text: prompt }],
        toolConfig,
        // default inference config
      }
    }
  }
}

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  const body = JSON.parse(ctx.result.body);
  const { content } = body.output.message;

  if (content.length < 1) {
    util.error('No content block in assistant response.', 'error');
  }

  const toolUse = content[0].toolUse;
  if (!toolUse) {
    util.error('Missing tool use block in assistant response.', 'error');
  }

  const response = toolUse.input.value;
  return response;
}"
`;

exports[`generation route model type with null timestamps 1`] = `
Object {
  "Fn::Join": Array [
    "",
    Array [
      "export const request = (ctx) => {
  ctx.stash.typeName = \\"Query\\";
  ctx.stash.fieldName = \\"makeTodo\\";
  ctx.stash.conditions = [];
  ctx.stash.metadata = {};
  ctx.stash.metadata.dataSourceType = \\"HTTP\\";
  ctx.stash.metadata.apiId = \\"",
      Object {
        "Fn::GetAtt": Array [
          "GraphQLAPI",
          "ApiId",
        ],
      },
      "\\";
  ctx.stash.connectionAttributes = {};
  ctx.stash.endpoint = \\"https://bedrock-runtime.",
      Object {
        "Ref": "AWS::Region",
      },
      ".amazonaws.com\\";
  ctx.stash.authRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "authRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.unauthRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "unauthRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.adminRoles = [];
  return {};
}

export const response = (ctx) => {
  return ctx.prev.result;
};",
    ],
  ],
}
`;

exports[`generation route model type with null timestamps 2`] = `
"export function request(ctx) {
    const toolConfig = {\\"tools\\":[{\\"toolSpec\\":{\\"name\\":\\"responseType\\",\\"description\\":\\"Generate a response type for the given field\\",\\"inputSchema\\":{\\"json\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"content\\":{\\"type\\":\\"string\\",\\"description\\":\\"A UTF-8 character sequence.\\"},\\"isDone\\":{\\"type\\":\\"boolean\\",\\"description\\":\\"A boolean value.\\"},\\"id\\":{\\"type\\":\\"string\\",\\"description\\":\\"A unique identifier for an object. This scalar is serialized like a String but isn't meant to be human-readable.\\"}},\\"required\\":[\\"id\\"]}},\\"required\\":[\\"value\\"]}}}}],\\"toolChoice\\":{\\"tool\\":{\\"name\\":\\"responseType\\"}}};
    const prompt = \`Make a string based on the description.\`;
    const args = JSON.stringify(ctx.args);

    return {
      resourcePath: '/model/anthropic.claude-3-haiku-20240307-v1:0/converse',
      method: 'POST',
      params: {
        headers: {
          'Content-Type': 'application/json',
        },
        body: {
          messages: [
          {
            role: 'user',
            content: [
              {
                text: args
              }
            ],
          }
        ],
        system: [{ text: prompt }],
        toolConfig,
        // default inference config
      }
    }
  }
}

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  const body = JSON.parse(ctx.result.body);
  const { content } = body.output.message;

  if (content.length < 1) {
    util.error('No content block in assistant response.', 'error');
  }

  const toolUse = content[0].toolUse;
  if (!toolUse) {
    util.error('Missing tool use block in assistant response.', 'error');
  }

  const response = toolUse.input.value;
  return response;
}"
`;

exports[`generation route scalar type 1`] = `
Object {
  "Fn::Join": Array [
    "",
    Array [
      "export const request = (ctx) => {
  ctx.stash.typeName = \\"Query\\";
  ctx.stash.fieldName = \\"makeTodo\\";
  ctx.stash.conditions = [];
  ctx.stash.metadata = {};
  ctx.stash.metadata.dataSourceType = \\"HTTP\\";
  ctx.stash.metadata.apiId = \\"",
      Object {
        "Fn::GetAtt": Array [
          "GraphQLAPI",
          "ApiId",
        ],
      },
      "\\";
  ctx.stash.connectionAttributes = {};
  ctx.stash.endpoint = \\"https://bedrock-runtime.",
      Object {
        "Ref": "AWS::Region",
      },
      ".amazonaws.com\\";
  ctx.stash.authRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "authRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.unauthRole = \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "unauthRoleName",
      },
      "/CognitoIdentityCredentials\\";
  ctx.stash.adminRoles = [];
  return {};
}

export const response = (ctx) => {
  return ctx.prev.result;
};",
    ],
  ],
}
`;

exports[`generation route scalar type 2`] = `
"export function request(ctx) {
    const toolConfig = {\\"tools\\":[{\\"toolSpec\\":{\\"name\\":\\"responseType\\",\\"description\\":\\"Generate a response type for the given field\\",\\"inputSchema\\":{\\"json\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"string\\",\\"description\\":\\"A UTF-8 character sequence.\\"}},\\"required\\":[\\"value\\"]}}}}],\\"toolChoice\\":{\\"tool\\":{\\"name\\":\\"responseType\\"}}};
    const prompt = \`Make a string based on the description.\`;
    const args = JSON.stringify(ctx.args);

    return {
      resourcePath: '/model/anthropic.claude-3-haiku-20240307-v1:0/converse',
      method: 'POST',
      params: {
        headers: {
          'Content-Type': 'application/json',
        },
        body: {
          messages: [
          {
            role: 'user',
            content: [
              {
                text: args
              }
            ],
          }
        ],
        system: [{ text: prompt }],
        toolConfig,
        // default inference config
      }
    }
  }
}

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  const body = JSON.parse(ctx.result.body);
  const { content } = body.output.message;

  if (content.length < 1) {
    util.error('No content block in assistant response.', 'error');
  }

  const toolUse = content[0].toolUse;
  if (!toolUse) {
    util.error('Missing tool use block in assistant response.', 'error');
  }

  const response = toolUse.input.value;
  return response;
}"
`;

exports[`generation route scalar type 3`] = `
"type Query {
  makeTodo(description: String!): String
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}
"
`;
